### PR TITLE
improve VCS URL detection and add opt-out

### PR DESCRIPTION
**Pull request by kaniini**
Address concerns based on over-the-weekend testing of the VCS URL feature:

- sanitize authentication data from the generated VCS URLs
- add ability to disable VCS URL probing outside of setting one explicitly in the config

### CODE REVIEW

File: internal/cli/publish.go [hunk 1]
```diff
@@ -82,6 +84,7 @@ in a keychain.`,
 	cmd.Flags().BoolVar(&useProot, "use-proot", false, "use proot to simulate privileged operations")
 	cmd.Flags().BoolVar(&useDockerMediaTypes, "use-docker-mediatypes", false, "use Docker mediatypes for image layers/manifest")
 	cmd.Flags().BoolVar(&debugEnabled, "debug", false, "enable debug logging")
+	cmd.Flags().BoolVar(&withoutVCS, "without-vcs", false, "disable VCS URL detection")
```
Comment (puerco) [inline internal/cli/publish.go:87]:
```suggestion
	cmd.Flags().BoolVar(&withVCS, "vcs", true, "detect and embed VCS URLs")
```

File: internal/cli/publish.go [hunk 2]
```diff
@@ -43,6 +43,7 @@ func publish() *cobra.Command {
 	var extraKeys []string
 	var extraRepos []string
 	var debugEnabled bool
+	var withoutVCS bool
```
Comment (puerco) [inline internal/cli/publish.go:46]:
```suggestion
	var withVCS bool
```

File: internal/cli/build.go [hunk 3]
```diff
@@ -80,13 +81,15 @@ bill of materials) describing the image contents.
...
 				build.WithDebugLogging(debugEnabled),
+				build.WithoutVCS(withoutVCS),
 			)
 		},
 	}
 
 	cmd.Flags().BoolVar(&useProot, "use-proot", false, "use proot to simulate privileged operations")
 	cmd.Flags().BoolVar(&useDockerMediaTypes, "use-docker-mediatypes", false, "use Docker mediatypes for image layers/manifest")
 	cmd.Flags().BoolVar(&debugEnabled, "debug", false, "enable debug logging")
+	cmd.Flags().BoolVar(&withoutVCS, "without-vcs", false, "disable VCS URL detection")
```
Comment (puerco) [inline internal/cli/build.go:92]:
```suggestion
	cmd.Flags().BoolVar(&withVCS, "vcs", true, "detect and embed VCS URLs")
```

File: internal/cli/build.go [hunk 4]
```diff
@@ -33,6 +33,7 @@ func buildCmd() *cobra.Command {
 	var useProot bool
 	var useDockerMediaTypes bool
 	var debugEnabled bool
+	var withoutVCS bool
```
Comment (puerco) [inline internal/cli/build.go:36]:
```suggestion
	var withVCS bool
```

File: internal/cli/build.go [hunk 5]
```diff
@@ -80,13 +81,15 @@ bill of materials) describing the image contents.
 				build.WithExtraRepos(extraRepos),
 				build.WithArch(types.ParseArchitecture(ba)),
 				build.WithDebugLogging(debugEnabled),
+				build.WithoutVCS(withoutVCS),
```
Comment (puerco) [inline internal/cli/build.go:84]:
```suggestion
				build.WithVCS(withVCS),
```

File: internal/cli/publish.go [hunk 6]
```diff
@@ -71,6 +72,7 @@ in a keychain.`,
 				build.WithExtraKeys(extraKeys),
 				build.WithExtraRepos(extraRepos),
 				build.WithDebugLogging(debugEnabled),
+				build.WithoutVCS(withoutVCS),
```
Comment (puerco) [inline internal/cli/publish.go:75]:
```suggestion
				build.WithVCS(withVCS),
```

File: pkg/build/build.go [hunk 7]
```diff
@@ -155,6 +156,10 @@ func New(workDir string, opts ...Option) (*Context, error) {
 		bc.Options.Arch = types.ParseArchitecture(runtime.GOARCH)
 	}
 
+	if !bc.Options.WithoutVCS && bc.ImageConfiguration.VCSUrl == "" {
```
Comment (puerco) [inline pkg/build/build.go:159]:
```suggestion
	if bc.Options.WithVCS && bc.ImageConfiguration.VCSUrl == "" {
```

File: pkg/build/options.go [hunk 8]
```diff
@@ -162,3 +164,11 @@ func WithDebugLogging(enable bool) Option {
 		return nil
 	}
 }
+
+// WithoutVCS disables VCS URL probing for the build context.
+func WithoutVCS(enable bool) Option {
+	return func(bc *Context) error {
+		bc.Options.WithoutVCS = enable
```
Comment (puerco) [inline pkg/build/options.go:171]:
```suggestion
// WithVCS enables VCS URL probing for the build context.
func WithVCS(enable bool) Option {
	return func(bc *Context) error {
		bc.Options.WithVCS = enable
```

File: pkg/options/options.go [hunk 9]
```diff
@@ -29,6 +29,7 @@ type Options struct {
 	UseDockerMediaTypes bool
 	WantSBOM            bool
 	UseProot            bool
+	WithoutVCS          bool
```
Comment (puerco) [inline pkg/options/options.go:32]:
```suggestion
	WithVCS          bool
```

**Review by puerco — COMMENTED** [review]
I personally find negative flags not that intuitive. What do you think about having a `--vcs` flag defaulting to `true` instead? To me, it feels more natural defining `--vcs` / `--vcs=false`.

Committing these suggestions should do it, but feel free to dismiss them if you prefer the without- → false way.

**Comment by kaniini** [comment]
Sure, what do you think now @puerco?

**Review by puerco — APPROVED** [review]
Nice, thank you!!

**Pull request merged**